### DISCUSSION TITLE
What is the idiomatic way in cfs for scheduling child tasks in an app? (Is it OS_TaskDelay?)

### DISCUSSION THREAD

**Discussion by embeddedpenguin** [nasa/cFS#817] [Q&A] [answered]
Looking at the current tasking example here:
https://github.com/nasa/osal/blob/main/src/examples/tasking-example/tasking-example.c
It looks like osal currently wants me to create my task and then stick a delay(ms) call at the bottom. Is there a way to schedule a child task to run at some frequency instead of using the delay method? Or is the delay method the only way of doing this currently?

Something like "I want to read this data 5x a second"?

**Comment by RichLandau** [comment]
Generally you'll use the scheduler (SCH) to wake up tasks at your desired rate, and have your task block on the software bus receive message call

**Reply by embeddedpenguin** [reply]
Are there any examples of this?

**Reply by RichLandau** [reply, marked as answer]
Not directly as far as I know, but you can see examples of:
1) [TIME service creating a child task](https://github.com/nasa/cFE/blob/main/modules/time/fsw/src/cfe_time_task.c#L165)
2) [Task blocking waiting to receive data from SB](https://github.com/nasa/sample_app/blob/main/fsw/src/sample_app.c#L77)
3) [SCH Lab configured to periodically send the message to wake it up](https://github.com/nasa/sch_lab/blob/main/fsw/tables/sch_lab_table.c#L84)

**Reply by embeddedpenguin** [reply]
Thanks!

**Comment by skliper** [comment]
You can also use a timer directly to run a loop at whatever frequency you want.  As an example, see https://github.com/nasa/sch_lab for setting up a timer that calls SCH_LAB_LocalTimerCallback, giving a semaphore to run the processing loop.
